typing/typecore.ml

```diff
@@ -2714,7 +2715,27 @@ let remaining_function_type ty_ret mode_ret rev_args =
   in
   ty_ret
 
-let collect_unknown_apply_args env funct ty_fun mode_fun rev_args sargs ret_tvar =
+let check_local_application_complete ~env ~app_loc ~sarg ~rest ~mode_fun ~mode_arg ~mode_ret =
```

Comment:
May I order a comment, please? Here is a start:

```
(* This function checks whether an application to local arguments is complete.
   It is necessary to avoid weird types being inferred in places where e.g. there is
   no mli file. For example, TODO should get the type TODO, not TODO.
   The error this produces is somewhat obscure, so prefer other errors over this one
   when there is a choice. This check could be dropped -- there's no soundness issue here --
   at the cost of a worse user interface. *)
```

I worry that, without such a comment, we'll be scratching our heads over this check in a few years.

Reply:
Sure, but I think this check goes away because we will be assuming that an application to n arguments is for an n parameter function, which will already force the desired property here. (We will still allow over-application when the function type is know, but that is also fine from this perspective).

This check is essentially adding mode constraints to enforce "if the function or a parameter is local then function should be at least arity n", whereas later we will just always enforce that the arity regardless of the modes of anything.